frontend/src/widgets/CoursesAssociatedWithActivityReports.js

```diff
@@ -0,0 +1,230 @@
+      sort.activePage = pageNumber;
+
+      // store it
+      setSortConfig(sort);
+      setOffset((pageNumber - 1) * perPage);
+    }
+  };
+
+  const requestSort = (sortBy) => {
+    // Get sort direction.
+    let direction = 'asc';
+    if (
+      sortConfig
+      && sortConfig.sortBy === sortBy
+      && sortConfig.direction === 'asc'
+    ) {
+      direction = 'desc';
+    }
+
+    // Set the value we want to sort by.
+    const valuesToSort = sortBy === 'Course'
+      ? courseUse.map((t) => ({
+        ...t,
+        sortBy: t.heading,
+      }))
+      : courseUse.map((t) => (
+        {
+          ...t,
+          sortBy: parseValue(t.data.find((tp) => tp.title === sortBy).value),
+        }));
+
+    // Value sort.
+    const sortValueA = direction === 'asc' ? 1 : -1;
+    const sortValueB = direction === 'asc' ? -1 : 1;
+    valuesToSort.sort((a, b) => {
+      if (a.sortBy > b.sortBy) {
+        return sortValueA;
+      } if (b.sortBy > a.sortBy) {
+        return sortValueB;
+      }
+      return 0;
+    });
+
+    setCourseUse(valuesToSort);
+    setOffset(0);
+    setSortConfig({ sortBy, direction, activePage: 1 });
+  };
+
+  const exportRows = (exportType) => {
+    let url = null;
+    try {
+      let coursesToExport = courseUse;
+      if (exportType === 'selected') {
+      // Get all the ids of the rowsToExport that have a value of true.
+        const selectedRowsStrings = Object.keys(checkBoxes).filter((key) => checkBoxes[key]);
+        // Loop all selected rows and parseInt to an array of integers.
+        const selectedRowsIds = selectedRowsStrings.map((s) => parseInt(s, DECIMAL_BASE));
+        // Filter the courses to export to only include the selected rows.
+        coursesToExport = courseUse.filter((row) => selectedRowsIds.includes(row.id));
+      }
+
+      // Create a header row.
+      const headerData = data.headers.map((h) => ({ title: h, value: h }));
+      headerData.push({ title: 'Total', value: 'Total' });
+      coursesToExport = [
+        {
+          heading: 'Course',
+          data: headerData,
+        },
+        ...coursesToExport,
+      ];
+
+      // create a csv file of all the rows.
+      const csvRows = coursesToExport.map((row) => {
+        const rowValues = row.data.map((d) => d.value);
+        // If the heading has a comma, wrap it in quotes.
+        const rowHeadingToUse = row.heading.includes(',') ? `"${row.heading}"` : row.heading;
+        return `${rowHeadingToUse},${rowValues.join(',')}`;
+      });
+      // Create CSV.
+      const csvString = csvRows.join('\n');
+      const blob = new Blob([csvString], { type: 'text/csv' });
+
+      // Check if url exists with the attribute of download.
+      if (document.getElementsByName('download').length > 0) {
+        document.getElementsByName('download')[0].remove();
+      }
+      url = window.URL.createObjectURL(blob);
+      const a = document.createElement('a');
+      a.setAttribute('name', 'download');
```

Comment:
name is not a valid attribute for <a>
You can use document.querySelector('[download="courses.csv"]') or something like that in your test

Reply:
Removed and updated.